Ingestion Service Charter

tburdett opened the pull request:
~**October 8: Last call for community reviews**~
~**October 16: Last call for oversight reviews**~
**October 18:** Approved on Architecture all

This PR is for the first draft of the Ingestion Service charter. Given ingest is quite complex with many moving parts, there is intentionally only high level data about the broad functionality of ingest, but little implementation details about the individual components. Detailed specifications will instead be provided in each component's repository. An overview will be made available in ingest-central.

Feedback is appreciated on both the content and the level of granularity, especially over the level of detail in milestones, which correspond to major anticipated ingest release

Commit:
Add ingest-infrastructure charter

This commit introduces the complete first draft of the ingest
infrastructure charter. Given ingest is quite complex with many moving
parts, there is intentionally only high level data about the broad
functional components of ingest, but little implementation details about
the individual components. Detailed specifications which instead will be
provided in each component's repository. An overview will be made
available in ingest-central.

Note that the repository list is still somewhat in flux and needs
consolidation and cleanup.

brianraymor committed:
Updates to section headings

The charter template was updated - #18

Commit:
Redraft for style consistency

Redrafting charter for writing style consistency and based on Matt W's 
feedback. This is now slightly cleaner and more terse.

Definitions that were previously scattered throughout the text have now 
been moved to a dedicated "Definitions" section at the end

I have left the subservice-based separation of scoping; collapsing into 
a single section makes it too complex and unreadable, but I have 
clarified in the description the rationale for this constituent service 
separation

Commit:
Rename 'Ingest Infrastructure' to 'Ingestion Service'

Renaming throughout to refer to the 'Ingestion Service', as this is a 
better description of what this charter really describes, but is not 
compatible with Brian R's original slides. I'm pretty sure this is 
acceptable though ;)

tburdett commented:
These changes should address most of your comments @mweiden - if you can review and check that'd be awesome!

I have left the "component" separation, I think given the diversity of stuff ingest has to do grouping around goals makes it almost unreadably complex. I've added some text to clarify the fact that this is structured around user profiles and goals rather than implementation practices. Feedback on this very welcome

tburdett commented:
One discussion point: I have cleaned up some definitions that were littered around into a dedicated section at the end. I feel like it's important to put some context to some of the jargon terms we use. Do others feel like this is a good idea per charter or DCP wide? I'd be very in favour of DCP wide but this will also incur some overhead. @brianraymor it'd be great to get your opnion on this!

mweiden commented on charters/ingest-infrastructure/charter.md:
```diff
@@ -0,0 +1,166 @@
+  * Interact with the DCP Upload Service to track and validate uploaded data files
+  * Track submission states that describe the life-cycle of experimental data as it moves through the DCP
+  * Generate identifiers for data, metadata and submissions
+  * Store of submitted data in the DCP Data Storage Service
```
~of~ ... or is that used in the Queen's English?

tburdett replied:
Nope, this is just a typo, thanks

brianraymor commented:
Regarding **jargon** in charters, @TimothyTickle posed a similar question for secondary analysis. As reviewers, we've been ensuring that everyone adopts the updated terms shared by Jodi - https://humancellatlas.slack.com/archives/C53S0GBS7/p1534526020000100 - @jodihir - is there a glossary on the data portal that charters could reference/link? And if so, does it define all the terms included in Tony's **Definitions** section?

mweiden reviewed (commented):
I like this new revision.

The one thing I'm a little stuck on is the framing of `In-scope` and `Out-of-scope` sections in terms of systems rather than thematic goals. Couple of thoughts here:

* The goals motivate what systems we build right? Feasibly, you could design a whole new set of systems that satisfy the same requirements.
* If you decided to replace a system--say you decided to throw out spreadsheets and make everyone use pandas dataframes--then you'd have to change the charter even though the goals of automating data-entry toil and uploading that data are essentially the same.

This is more of a nit though, not a firm ask.

Commit:
Fix 'store of' typo

tburdett commented:
Yeah, so I'm sympathetic to the idea of aiming at goals rather than systems. However, the sub-systems are so tied to user profiles and requirements that it's hard to separate them out.

Spreadsheets: we're building systems around them because all of our submitters tell us they just want to give us spreadsheets. At this point the template generation and spreadsheet upload process becomes a requirement rather than an implementation choice. I did try to rejig the scoping sections around goals, but I ended up with the same goals being delivered four different ways for the different user profiles and it became very difficult to read. If others feel strongly I'll make a more rigorous attempt :smile:

Commit:
Fix 'a spreadsheets' typo

Commit:
Fix capitalisation on 'Spreadsheet Template generator'

jodihir commented:
In response to @brianraymor's question, there is not currently a glossary on the portal. There are a couple of glossary-type docs that I know of in the DCP Google drive ; maybe next quarter we could combine these, add Tony's terms, and start to create a page for the portal. I think as we get all the content into the portal we'll be able to figure out what else should go in a glossary for users.  But would we want a separate glossary for internal use? https://docs.google.com/document/d/1DmSoUxfblKMZTb7fkI7mQ0a4dHHXpTTNJ4Ho7Fm9hDY/edit
https://docs.google.com/spreadsheets/d/1yOjsfZptQJiuagVLpy4xlNnL9-yKK87gTs5gWIare2c/edit#gid=0

brianraymor commented:
Based on @jodihir's [response](https://github.com/HumanCellAtlas/dcp-community/pull/16#issuecomment-416676736), perhaps the simplest approach would be to add an optional **Definitions** section to the charter template between **Description** and **Objectives** to address emerging requirements from both @tburdett and @TimothyTickle.  We can assess whether we want to create a single glossary that can be referenced by all charters later. Comments?

tburdett commented:
@brianraymor I'm very happy with an optional Definitions section as a pragmatic solution and to let a DCP wide glossary bubble out of the terms that are used consistently :+1:

Commit:
Move 'Definitions' section

'Definitions' now occurs between 'Description' and 'Objectives'

lauraclarke commented:
A couple of questions about your definitions @tburdett

`Data wrangler - a data contributor who works as part of the HCA consortium to provide assistance to biologists wanting to upload data into the DCP`

The data wrangler definition not using the word data contributor for who they are supporting is a little concerning. The labs to submit data to us are data contributors even if they are getting support from data wranglers, I wonder if calling data wranglers data contributors (I assume this is to avoid having to say data contributors and data wranglers when referring to interactions with the systems) is being overly precise

`Submission - the grouping of data and metadata that makes sense to a data contributor. For example, a data contributor might typically upload all the data at once for a publication they have in preparation. This grouping does not necessarily correspond to how data is stored or displayed in the DCP and how it should be displayed to a data consumer`

For submission, I think you are missing the important aspect of to submit at the same time, in a single operation with the system (via wranglers or not). The logical grouping of data from a dataset perspective is a project. A project may be made up of one or more submissions. That seems to be lost in this definition

Commit:
Rename ingest-infrastructure directory to ingestion-service

ingest-infrastructure directory has been renamed to match the renaming 
of the charter to ingestion-service

Commit:
Update definitions based on Laura's feedback

Minor introduction update and work on definitions to address comments.

Data wrangler revisions:

 - Consistency with Ingest Support charter
 - Better overall definition
 - Illustratation that they can be considered as a type of contributor 
from the Ingestion Service perspective only
 
Submission revisions:

 - Clarifies that submissions occur as a single payload

lauraclarke commented on charters/ingest-infrastructure/charter.md:
```diff
@@ -0,0 +1,166 @@
+
+*Direct data contributor* - a data contributor who uploads their own data through services offered as part of the DCP in a "self-service" manner. Direct data contributors are also usually the generators of the data
+
+*Data wrangler* - a data contributor who works as part of the HCA consortium to provide assistance to biologists wanting to upload data into the DCP
```
Could this be changed to 

Data wrangler - the Ingest Support team member who assists labs contributing data to the HCA. In the context of this charter, data wranglers considered as data contributors

or something along these lines

It is important to note that data wranglers will also support the direct data contributors in your parlance but they won't do as much for them.

briandoconnor commented on charters/ingestion-service/charter.md:
```diff
@@ -0,0 +1,166 @@
+ * Dedicated views over the data, for example visualisation of all data for a particular project or from a particular tissue type
+
+### Spreadsheet Template Generator
+ * Generation of lab specific spreadsheets that are not directly generated from a JSON schema endorsed by the metadata team. Additional customisations (for example, modifying column names or adding additional tabs) for individual contributors will not be supported through any mechanism other than through the use of standard configuration (i.e. which modules are to be included) or through the prop ...
```
One comment I have is how does Ingest deal with data that is either 1) not well supported and bleeding edge and 2) has metadata specifications but is currently not being processed by Green.  Seems like 2 should definitely but uploadable but not sure about 1.  Is there a path to submitting data (with, say, minimal metadata) to get it in blue and available in orange even before the full metadata model is not yet prepared?  This overlaps with the metadata charter too but I'd like to hear about it from the perspective of ingest.

tburdett replied:
Yes, great question. The ingest perspective is that anything with a defined metadata schema that is a valid JSON schema can be uploaded and will be validated, and if valid will be deposited in the DSS. I think this covers your scenario 2.

It is also possible, if you have truly bleeding edge data (scenario 1), to add additional metadata fields (e.g. creating new columns in the spreadsheet) to push data to ingest. This won't validate against the current schema (and therefore won't be released to the DSS), but we have provision to revalidate once a new schema is published. Of course, the alternative is that you create a submission that validates against the last schema but e.g. misses some new fields needed, and then providing an update once the new schema is released. Updates need to be supported before we can deliver on this but this is the plan.

Hopefully I've explained that clearly? Most of this functionality is already supported in the spreadsheet template importer (@simonjupp can provide implementation details if you're interested), and for programmatic submissions we *always* retain whatever is sent even if it's invalid

tburdett replied:
Do you have a suggestion for how I can update the charter to help communicate this better?

kbergin commented:
Hello! 

Would you like to add a reference to the new upload team email on this charter? 

On the Data Processing Charter I decided to make the title of the charter `mailto:email@data.humancellatlas.org` and to add it to the Communication section. I formatted it as
`[Team email](mailto:email@data.humancellatlas.org): email@data.humancellatlas.org`

For your reference, your team email is `ingest-team@data.humancellatlas.org`

Thanks! I'll be posting this on each charter for convenience, please do let me know your thoughts :)

Commit:
Reorganise content for consistency with Data Store Charter

This commit reorganises the scoping sections of the Ingestion Service 
charter to be more consistent with the current Data Store charter. It 
focuses on a separation between interfaces and core capabilites and 
avoids very heavily component based headings as suggested by @mweiden.

Commit:
Add h1 email address from @kbergin's suggestion

mweiden commented on charters/ingestion-service/charter.md:
```diff
@@ -0,0 +1,173 @@
+
+# [Ingestion Service](mailto:ingest-team@data.humancellatlas.org)
```
Why not add the email to the Communication section as Kylee suggested?

brianraymor committed:
Added mailing list to Communications section

kbergin commented:
To be fair I also suggested making my header a link to an email. Open to
not doing that as well.

On Thu, Oct 4, 2018 at 6:33 PM Matt Weiden <notifications@github.com> wrote:

> *@mweiden* commented on this pull request.
> ------------------------------
>
> In charters/ingestion-service/charter.md
> <https://github.com/HumanCellAtlas/dcp-community/pull/16#discussion_r222847663>
> :
>
> > @@ -0,0 +1,173 @@
> +
> +# [Ingestion Service](mailto:ingest-team@data.humancellatlas.org)
>
> Why not add the email to the Communication section as Kylee suggested?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/HumanCellAtlas/dcp-community/pull/16#pullrequestreview-161835777>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJAxJMG7QE0y155XgSyZh8hKPIz3kUfvks5uhozTgaJpZM4WESWD>
> .
>
-- 
Kylee Degatano
Senior Software Product Manager
Human Cell Atlas Data Coordination Platform
Data Sciences Platform, Broad Institute

brianraymor commented:
@kbergin  and @mweiden  - The charter template uses the **Communication** section to:

_List the Slack Channels, Mailing Lists, and/or Discussion Forums for the Project and describe their purpose._

It's fine to also link the Project Name although I'm uncertain about the benefit - but the priority should be adding mailing lists to  **Communication**. Looking forward, I'm wondering whether we can automate the creation of some community contact pages by iterating over the markdown.

malloryfreeberg commented on charters/ingestion-service/charter.md:
```diff
@@ -0,0 +1,177 @@
+The secondary mission is to provide convenient, user-friendly methods for the upload of data and metadata to the DCP. The Ingestion Service will provide user interfaces and tools suitable for a variety of data contributor profiles, and will continuously review whether these interfaces and tools adequately serve the needs of the HCA user community. As part of this secondary mission, the Ingestion  ...
+
+  * A **User Interface** targeting data contributors uploading data and metadata via a form-based user interface
+  * A **Spreadsheet Upload** service targeting data generators uploading experimental metadata using spreadsheets in single, canonical defined spreadsheet format based on a specific Excel template (.xlsx) created by the spreadsheet template generator
```
> based on a specific Excel template (.xlsx)

I'm worried about locking the Ingest Service in to a single spreadsheet implementation format here (although I appreciate that this is the only format the Ingestion Service supports at the moment).

tburdett replied:
Given that we make extensive use of tabs to represent different metadata sheets, the common alternatives (CSV, TSV) aren't appropriate. OpenDocument (.ods) format might be, but I think this will be more work than benefit (I've never seen a biologist save a spreadsheet in Excel using anything other than .xslx). Supporting Google Sheets seems like a very good idea, though. I think that covers the whole landscape though.

Summarising, I think the options are to to remove the "Excel" part, or add additional text, maybe something like

> "...based on a specific template available for Excel (.xlsx) or Google Sheets..."

Removing any reference seems a bit enigmatic given we're talking about a very small number of routes that will be supported. Would you prefer one over the other? Or something else?

malloryfreeberg replied:
I mean, someone might be using old .xls Excel format :)

I'm happy to just have Excel and Google Sheets mentioned.

malloryfreeberg commented on charters/ingestion-service/charter.md:
```diff
@@ -0,0 +1,177 @@
+* Provide state tracking services that supply information about the current state of any given submission, allowing submitters to see where in the ingestion process their data is currently
+* Provide services for connecting metadata describing experimentally generated data files, and the locations of those data files in the cloud
+* Provide services that syntactically validate experimentally generated data files, covering commonly used and endorsed community data formats (e.g. FASTQ)
+* Provide services that allow data contributors to confirm their submissions, prompting storage data and metadata in the DCP data store
```
> prompting storage data and metadata in the DCP data store

prompting storage **of** data and metadata in the DCP **Data Store**

malloryfreeberg commented on charters/ingestion-service/charter.md:
```diff
@@ -0,0 +1,177 @@
+### Supporting tools
+
+* Provide tools that use reference (latest) metadata schemas, defined by the metadata team, to produce template spreadsheets in a canonical spreadsheet format, such that data contributors can easily complete this spreadsheet to capture their metadata
+* Provide tools that allow data contributors (normally, data wranglers) to generate customised spreadsheets that target specific experiment types or data types based on a small number of screening criteria
```
> (normally, data wranglers)

In the near future, this might be mostly data wranglers, but I don't think it's limited to data wranglers. The goal is for any data contributor to come in to the Ingestion Service and be given tools to generate their won customized template spreadsheet. I would remove this phrase.

malloryfreeberg commented on charters/ingestion-service/charter.md:
```diff
@@ -0,0 +1,177 @@
+
+### Core capabilities
+
+* Metadata specification definitions. Ingest will utilise metadata definitions defined by DCP-compliant JSON schemas, as defined by the metadata working group, and will ensure the Ingest API is schema-agnostic where possible, but will not actively define metadata specifications beyond the restrictions placed by the shared core domain model outlined above
```
> defined by the metadata working group

Since we don't really have a metadata working group, can this be changed to "defined by the metadata team"?

malloryfreeberg commented on charters/ingestion-service/charter.md:
```diff
@@ -0,0 +1,177 @@
+### Core capabilities
+
+* Metadata specification definitions. Ingest will utilise metadata definitions defined by DCP-compliant JSON schemas, as defined by the metadata working group, and will ensure the Ingest API is schema-agnostic where possible, but will not actively define metadata specifications beyond the restrictions placed by the shared core domain model outlined above
+* Definition of bundle structures for the DCP datastore. Whilst the Ingest API will provide support for generating bundles in response to submissions, the Ingestion Service is not responsible for the specification of bundle structures - it is expected these will be provided either by the metadata team or to address specific consumer requirements
```
I don't think bundle structure specifications are provided by the metadata team. I thought they were more constrained by the DS (to meet indexing requirements) and/or the Pipelines team (to meet analysis requirements).

tburdett replied:
I'm going to resolve this, as this was discussed quite a lot amongst PMs and POs. The Data Store charter now says this:
> The Data Store team will own the process of collecting new bundle requirements from users and assessing them against the datastore design before sending them to the metadata team for the definition of a new spec. The Data Store owns the bundle use cases and bundle type definitions, while the precise specifications will be negotiated between the Metadata and other teams.

... and the Metadata Charter currently says this:
> Working with Ingestion Service to specify the entity structure for HCA metadata schema

It might be worth updating the Metadata Charter to reflect the definition of the bundle specification, which is really just a mapping of which entities go into which bundle type, as defined by the DSS. @lauraclarke do you have thoughts?

I'll add a suggestion to the Metadata Charter reflecting this comment

lauraclarke replied:
This is done

malloryfreeberg commented on charters/ingestion-service/charter.md:
```diff
@@ -0,0 +1,177 @@
+
+### Technical Lead
+
+[Simon Jupp](mailto:jupp@ebi.ac.uk) (Interim, pending new hire)
```
Can we add the new hire here yet?

tburdett replied:
I'll update the charter once it is official

malloryfreeberg commented on charters/ingestion-service/charter.md:
```diff
@@ -0,0 +1,177 @@
+
+[HumanCellAtlas #ingestion-service](https://humancellatlas.slack.com/messages/ingestion-service) for all discussions on the Ingestion Service
+
+[HumanCellAtlas #dcp-reuse](https://humancellatlas.slack.com/messages/dcp-reuse) if you are interested in using ingest, or other DCP components, in your own projects
```
> in using ingest,

Maybe change to "in using the Ingestion Service," to be consistent.

Commit:
Fix a typo spotted by @malloryfreeberg

This addresses Mallory's comment spotting the *of* typo, 
https://github.com/HumanCellAtlas/dcp-community/pull/16#discussion_r222936448

Commit:
Remove 'data wranglers' reference

Addresses @malloryfreeberg's comment on data wranglers, 
https://github.com/HumanCellAtlas/dcp-community/pull/16#discussion_r222937260

Commit:
Replace 'metadata working group' with 'metadata team'

Addresses @malloryfreeberg's comment, 
https://github.com/HumanCellAtlas/dcp-community/pull/16#discussion_r222938636

Commit:
Apply the use of "Ingestion Service" consistently

Addresses @malloryfreeberg's comment on the use of ingest vs Ingestion 
Service 
https://github.com/HumanCellAtlas/dcp-community/pull/16#discussion_r222947168

brianraymor committed:
Editorial updates for Slack Channels

Adopt best practice for slack channel format

Commit:
Rename ingestion-service directory to IngestionService

Renaming charter subdirectory for consistency with other charters

Commit:
Rephrase ungainly 'user interface' wording

mweiden commented:
Approved in the 2018-10-18 Architecture Meeting.
